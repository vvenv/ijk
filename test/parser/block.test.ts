import { describe, expect, test } from 'vitest';
import { parse } from './__helper';

describe('validation', () => {
  test('super() should only be a child of block', () => {
    expect(() => parse('{{ if x }}{{ super() }}{{ endif }}')).toMatchSnapshot();
    expect(() =>
      parse('{{ if x }}{{ super() }}{{ endif }}', { debug: true }),
    ).toThrowErrorMatchingSnapshot();
    expect(() => {
      try {
        parse('{{ if x }}{{ super() }}{{ endif }}', { debug: true });
      } catch (error: any) {
        throw new Error(error.details);
      }
    }).toThrowErrorMatchingSnapshot();
  });
});

test('basic', () => {
  expect(
    parse(
      `{{ block title }}1{{ endblock }}{{ block title }}2{{ endblock }}{{ block title }}{{ super() }}3{{ endblock }}`,
    ),
  ).toMatchSnapshot();
});

test('block tag should be at the root level', () => {
  expect(() =>
    parse('{{ if x }}{{ block title }}{{ endblock }}{{ endif }}', {
      debug: true,
    }),
  ).toThrowErrorMatchingInlineSnapshot(
    `[ASTError: "block" must follow "root", not "if".]`,
  );
});
