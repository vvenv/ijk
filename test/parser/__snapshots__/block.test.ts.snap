// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`basic 1`] = `
{
  "out": Out {
    "content": ""use strict";let s = "";const _b_1_0_0 = (_s)=>{s+="1";};const _b_1_0_32 = (_s)=>{s+="2";};const _b_1_0_64 = (_s)=>{_s?.();s+="3";};(() => _b_1_0_64(() => _b_1_0_32(() => _b_1_0_0())))();return s;",
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
  },
  "smp": SMP {
    "options": {
      "autoEscape": true,
      "debug": false,
      "strictMode": true,
      "stripComments": true,
      "tagEnd": "}}",
      "tagStart": "{{",
      "trimWhitespace": true,
    },
    "template": "{{ block title }}1{{ endblock }}{{ block title }}2{{ endblock }}{{ block title }}{{ super() }}3{{ endblock }}",
  },
}
`;

exports[`validation > super() should only be a child of block 1`] = `[Function]`;

exports[`validation > super() should only be a child of block 2`] = `[ASTError: "super" must follow "block", not "if".]`;

exports[`validation > super() should only be a child of block 3`] = `
[Error:  i>j>k  "super" must follow "block", not "if".

1: {{ if x }}{{ super() }}{{ endif }}
   ^^^^^^^^^^
             ^^^^^^^^^^^^^
]
`;
